### ISSUE TITLE
selector-max-compound-selectors

### ISSUE THREAD

**Issue by dryoma** [stylelint/stylelint#1141]
I'd like to propose a feature to consider: **selector-depth**

There is `max-nesting-depth` that controls the number of levels in Sass nesting. At first I thought it controls selector depth as well, but found that for regular selectors there is `selector-max-specificity`. That helps but not in all the cases.

Consider the case when we want maximum 3 levels of selector depth and don't want to limit selectors by their type (id/class/etc.). It is natural for us to want this:
```
.class1 .class2 .class3 { ... } // 0. We want this to work.
.class1 article .class3 { ... } // 1. And this.

.menu .item div a { ... } // 2. But not this.
.menu li div a span span { ... } // 3. And not this.
ul li ul li div a span span { ... } // 4. No way we would want this!
```
To allow the **0** we should set `selector-max-specificity: "0,3,0"`. Which will also allow 2, 3 and 4. The **3** illustrates the situation the best: it has specificity of *just 8* to forbid a single class/pseudo-element/etc. and *as much as 8* to allow eight levels of selector depth.

SCSS-lint [has this feature](https://github.com/brigade/scss-lint/blob/master/lib/scss_lint/linter/README.md#selectordepth) for a while now.

What do you think?

**Comment by jeddy3** [comment]
@dryoma Thanks for the suggestion and write up!

> What do you think?

I'm thinking this would be a good candidate for a [plugin](https://github.com/stylelint/stylelint/blob/master/docs/developer-guide/plugins.md) ("rules built by the community that support methodologies, toolsets, non-standard CSS features, *or very specific use cases*").

In the age of [BEM-like methodologies](https://suitcss.github.io/), and now [CSS Modules](https://github.com/css-modules/css-modules), I think this rule is only going get more edge-case as time goes on until it's eventually redundant.

The combination of (for CSS Modules):

- `selector-max-specificity`
- `selector-no-type`
- `selector-no-id`

Or using (for BEM-like stuff):

- `stylelint-selector-bem-pattern`

Cover the *majority* of use cases e.g.

```json
{
  "rules": {
    "selector-max-specificity": [0,1,1],
    "selector-no-type": true,
    "selector-no-id": true
  }
}
```

For CSS Modules forces you to only write:

```css
.className::after {}
.stateClassName {
  composes: className;
}
```

As an aside, you could even argue that `max-nesting-depth` is a bit redundant since the creation of `selector-max-specificity`, which can resolve nested selectors.

Anyhow, for the specific use-case of this:

```scss
.class1 .class2 .class3 { ... } // 0. We want this to work.
.class1 article .class3 { ... } // 1. And this.
.menu .item div a { ... } // 2. But not this.
```

I think it should be outside of core.

Thoughts?

**Comment by davidtheclark** [comment]
> As an aside, you could even argue that max-nesting-depth is a bit redundant since the creation of selector-max-specificity, which can resolve nested selectors.

I don't know — I feel like these are different ways to enforce a similar best practice. The new proposal here, `max-selector-combinators` or something, is similar: another approach to the same basic problem. I don't think we can say that `selector-max-specificity` is really the one-size-fits-all answer to this. I think it's legitimate to want a simpler (but less precise) rule like @dryoma proposes here.

Because I think this is probably a super simple rule, a lot of people would like it, and it is all about standard CSS, I wouldn't mind seeing it in core.

**Comment by dryoma** [comment]
> The combination of (for CSS Modules):
> Or using (for BEM-like stuff)

I personally like SUIT very much and am adopting `postcss-bem-linter` as soon as nesting is implemented there. But all those are either a conventions and not a standard, or future features adopting which could take browsers a long time. While, as @davidtheclark said, the proposal is about an old simple standard CSS.

Meanwhile I created a plugin for this: [stylelint-selector-max-depth](https://github.com/dryoma/stylelint-selector-max-depth/blob/master/README.md), feel free to examine and comment (sorry, having troubles registering it in NPM, so not making a PR to the plugin list yet).

**Comment by evilebottnawi** [comment]
https://github.com/stylelint/stylelint/pull/1159

**Comment by evilebottnawi** [comment]
Done in https://github.com/stylelint/stylelint/pull/1167
